pages/dao/[symbol]/proposal/components/instructions/Mango/MangoV4/TokenAddBank.tsx

```diff
@@ -0,0 +1,174 @@
+import { serializeInstructionToBase64 } from '@solana/spl-governance'
+import { AccountType, AssetAccount } from '@utils/uiTypes/assets'
+import InstructionForm, {
+  InstructionInput,
+  InstructionInputType,
+} from '../../FormCreator'
+import UseMangoV4 from '../../../../../../../../hooks/useMangoV4'
+import { TOKEN_PROGRAM_ID } from '@solana/spl-token'
+
+type NamePkVal = {
+  name: string
+  value: PublicKey
+}
+
+interface TokenAddBankForm {
+  governedAccount: AssetAccount | null
+  token: null | NamePkVal
+}
+
+const TokenAddBank = ({
+  index,
+  governance,
+}: {
+  index: number
+  governance: ProgramAccount<Governance> | null
+}) => {
+  const wallet = useWalletStore((s) => s.current)
+  const { getClient, GROUP } = UseMangoV4()
+  const { realmInfo } = useRealm()
+  const { assetAccounts } = useGovernanceAssets()
+  const governedProgramAccounts = assetAccounts.filter(
+    (x) => x.type === AccountType.SOL
+  )
+  const [tokens, setTokens] = useState<NamePkVal[]>([])
+  const { connection } = useWalletStore()
+  const shouldBeGoverned = !!(index !== 0 && governance)
+  const programId: PublicKey | undefined = realmInfo?.programId
+  const [form, setForm] = useState<TokenAddBankForm>({
+    governedAccount: null,
+    token: null,
+  })
+  const [formErrors, setFormErrors] = useState({})
+  const { handleSetInstructions } = useContext(NewProposalContext)
+  const handleSetForm = ({ propertyName, value }) => {
+    setFormErrors({})
+    setForm({ ...form, [propertyName]: value })
+  }
+  const validateInstruction = async (): Promise<boolean> => {
+    const { isValid, validationErrors } = await isFormValid(schema, form)
+    setFormErrors(validationErrors)
+    return isValid
+  }
+  async function getInstruction(): Promise<UiInstruction> {
+    const isValid = await validateInstruction()
+    let serializedInstruction = ''
+    if (
+      isValid &&
+      programId &&
+      form.governedAccount?.governance?.account &&
+      wallet?.publicKey
+    ) {
+      const client = await getClient(connection, wallet)
+      const group = await client.getGroup(GROUP)
+      const token = group.banksMapByMint.get(form.token!.value.toBase58())![0]
+      const mintInfo = group.mintInfosMapByTokenIndex.get(token.tokenIndex)
+      const bank = group.banksMapByTokenIndex.get(token.tokenIndex)
+      const ix = await client.program.methods
+        .tokenAddBank(Number(token.tokenIndex), Number(token.bankNum))
+        .accounts({
+          group: group.publicKey,
+          admin: form.governedAccount.extensions.transferAddress,
+          mint: token.mint,
+          payer: form.governedAccount.extensions.transferAddress,
+          rent: SYSVAR_RENT_PUBKEY,
+          tokenProgram: TOKEN_PROGRAM_ID,
+          systemProgram: SYSTEM_PROGRAM_ID,
+          existingBank: PublicKey.default,
+          mintInfo: mintInfo!.publicKey,
+          vault: token.vault,
+          bank: bank![0].publicKey,
```

Comment:
Should it actually be get from findProgramaddress ? @microwavedcola1

Reply:
this would be a new bank - so a new pk, and the bank_num would usually be 1 higher than the max bank_num in all other banks for same mint

you can either use a findProgramaddress or let anchor auto figure it for you afaik